src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs

```diff
@@ -654,17 +660,23 @@ private bool CheckCopyToOutputDirectoryFiles(Log log, in TimestampCache timestam
         {
             string outputFullPath = Path.Combine(state.MSBuildProjectDirectory, state.OutputRelativeOrFullPath);
 
-            foreach ((_, ImmutableArray<UpToDateCheckInputItem> items) in state.InputSourceItemsByItemType)
+            foreach ((string itemType, ImmutableArray<UpToDateCheckInputItem> items) in state.InputSourceItemsByItemType)
             {
                 foreach (UpToDateCheckInputItem item in items)
                 {
-                    // Only consider items with CopyType of CopyIfNewer (PreserveNewest)
-                    if (item.CopyType != CopyType.CopyIfNewer)
+                    token.ThrowIfCancellationRequested();
+
+                    if (item.CopyType == CopyType.Never)
                     {
+                        // Ignore items which are never copied. Only process Always and PreserveNewest items.
+                        //
+                        // Note that if we see Always items, then state.IsTreatCopyAlwaysAsPreserveNewestDisabled must
+                        // be false, as when it is true and any Always item exists, the check returns before this point.
+
                         continue;
                     }
 
-                    token.ThrowIfCancellationRequested();
+                    System.Diagnostics.Debug.Assert(item.CopyType != CopyType.Always || !state.IsTreatCopyAlwaysAsPreserveNewestDisabled);
```

Comment:
Did you mean to leave this in?

Reply:
Yes, it's a valid assertion. Though I don't want to throw at runtime for the user should it ever not hold.